Transaction

no-reply opened the pull request:
For initial review. This gets changesets and transactions in a state that I think they are usable.

Before these changes, Transaction fails regularly with no method errors on nil  `#changes`. Instead of "talking to" a nil on `changes` we introduce `Changeset#empty?` and give each transaction a changeset on initialization.

This also implements `Transaction#execute`, pushing responsibility for transactionality down to the `Repository`, via `Changeset#apply`. `Enumerable#supports?(:transactions)` is introduced; and changesets can now be applied to any `Mutable`. `Transaction` is still a `Repository` only concept.

@bendiken, @gkellogg this could use an initial review. In the meanwhile, I'm picking up work on transactionality tests & support for the default repository.

no-reply commented:
Note that class level documentation needs a rework before this is merged.

no-reply commented:
A dumb transactional repository implementation lives at https://github.com/ruby-rdf/rdf/tree/feature/repository-transactions

I think some tests should be put in place before something with better memory efficiency happens.

gkellogg commented:
Looks good to me.

no-reply commented on lib/rdf/changeset.rb:
```diff
@@ -1,24 +1,48 @@
+  # @note When applying a Changeset, deletes are resolved before inserts.
   #
   # @since 2.0.0
   class Changeset
```
I think `Changeset` is solid, now. 

Note that `#apply_changeset` is a method on `Mutable`, so `Changeset`s can be applied more generally than just to repositories.

no-reply commented on lib/rdf/mixin/enumerable.rb:
```diff
@@ -70,6 +70,7 @@ module Enumerable
     #   * `:inference` supports RDFS inferrence of queryable contents.
     #   * `:validity` allows a concrete Enumerable implementation to indicate that it does or does not support valididty checking. By default implementations are assumed to support validity checking.
     #   * `:skolemize` supports [Skolemization](https://www.w3.org/wiki/BnodeSkolemization) of an `Enumerable`. Implementations supporting this feature must implement a `#skolemize` method, taking a base URI used for minting URIs for BNodes as stable identifiers and a `#deskolemize` method, also taking a base URI used for turning URIs having that prefix back into the same BNodes which were origin ...
+    #   * `:transactions` supports atomic application of a [Changeset] 
```
Note to myself to reword this.

no-reply commented on lib/rdf/repository.rb:
```diff
@@ -195,22 +197,18 @@ def transaction(graph_name = nil, &block)
     # @param  [RDF::Transaction] tx
     # @return [void] ignored
     # @since  0.3.0
     def rollback_transaction(tx)
```
Rollback is now a real thing, and up to the `Transaction` to implement. The base implementation is to simply clear `Transaction#changes`.

Clients are discouraged from trying to re-use rolled back transactions. e.g. for:

```ruby
tx = Transaction.new(repo)
tx.insert(st1)
tx.rollback
tx.insert(st2)
tx.execute
```

Some Transactions may handle this fine, but subclasses for individual repositories are not required to handle it. It's allowable to open a transaction with the upstream repository, kill it on rollback, and provide no mechanism to reopen (or open a new transaction) within the same `Transaction` instance.

If there are ideas about how to lock this down a bit better, I would be happy to hear them.

gkellogg replied:
Perhaps you can freeze the transaction object on rollback, which should (could) cause any attempt to reuse the transaction to fail.

no-reply replied:
This seems like a good thought. :+1:

no-reply replied:
I'm reminded (again) that `#freeze` doesn't freeze instance variables of the frozen object.

Considering just adding a `@rolledback` flag, and raising an error on `#execute`. If it's true. (could do that with `#frozen?`, too, but it feels like a semantic overload)

no-reply replied:
Okay. That's in.

no-reply commented on lib/rdf/transaction.rb:
```diff
@@ -2,72 +2,110 @@ module RDF
   #   end
+  # 
+  # The base class provides a full, buffered implementation depending on 
+  # `RDF::Changeset` and using `Changeset#apply`. Custom `Repositories`
```
This attempts to clearly document the easy road to `Changeset` driven transactions via the base class. Useful?

gkellogg commented on lib/rdf/transaction.rb:
```diff
@@ -2,72 +2,110 @@ module RDF
+  # can implement a minimial write-atomic transactions by overriding
+  # `#apply_changeset`. Minimal snapshot isolation for queries can be 
+  # implemented by passing a `:snapshot` to `options` on initialization.
+  #
```
Could `snapshot` transactions be used within `Repository#each_statement` and `Repository#query_pattern` to eliminate the need to duplicate objects? Or would this simply happen elsewhere?

no-reply replied:
I'm not sure I follow this question? Can you clarify what the proposal would be?

gkellogg replied:
A snapshot is immutable, as I understand it. `#each_statement` needs to dup each level of the g/s/p/o structure because an insert or delete done while processing could mess up the enumeration. If we wrapped a snapshot around it, presumably, there could be no modification, so the duplication would not be necessary.  But, there may be some details of the implementation which make this impractical or undesirable.

no-reply replied:
Ah, yes. I'm following now. Yes, I think that would be up to the `Repository` (i.e. to be implemented as part of the WIP in https://github.com/ruby-rdf/rdf/tree/feature/repository-transactions).

I am thinking that this would benefit from a refactor, though. Perhaps the client doesn't need to pass a snapshot in, but rather `Repository` can generate it.  So the initializer becomes something like:

```ruby
@snapshot = repo.supports(:snapshot) ? repo.snapshot : repo
```

no-reply commented on lib/rdf/transaction.rb:
```diff
@@ -2,72 +2,110 @@ module RDF
+
+    extend Forwardable
+
+    def_delegators :@snapshot, :query_pattern, :query_execute
```
`Queryable` support is broken because the `Queryable` tests (and some of the actual implementation) rely on `Enumerable` methods. 

I'm thinking this is a bad thing for this use case. It's a high bar to expect a fully enumerable snapshot; while queryable is a lower bar. Blazegraph, for example, gives us an easy snapshot query with its own transactions API, but I'm not sure whether it would be good or efficient (though it's probably possible) to offer `#each` against that API.

Thoughts?

gkellogg replied:
Queryable includes Enumerable, so it seems that it seems that it may be necessary. But, the reason for this may have more to do with default implementation, rather than an abstract need for queryable to also be enumerable. It would require some re-factoring of Queryable::Enumerator.

no-reply commented on lib/rdf/transaction.rb:
```diff
@@ -81,38 +119,48 @@ def context=(value)
+    def initialize(repository, options = {}, &block)
+      @repository = repository
+      @options  = options.dup
+      @snapshot = @options.delete(:snapshot) { @repository }
```
This interface just came together in the last hour. I'm somewhat uncertain about it, and open to suggestions for revision.

gkellogg replied:
It should probably use keyword arguments, at least for :mutable, and :snapshot, as well as double-splat for options. Keeping `@options` may not be important after that. You can also give keyword arguments default values directly.

    def initialize(repository, mutable: false, snapshot: nil)
      @repository = repository
      @mutable = mutable
      @snapshot = snapshot || repository
      @changes = RDF::Changeset.new
    end

no-reply replied:
Agreed. I realized that I was still working with @bendiken's pre keyword args version when adding this, but didn't want to conflate that change with the `:snapshot` one. I'll refactor for keyword arguments in a later commit.

no-reply commented on lib/rdf/transaction.rb:
```diff
@@ -81,38 +119,48 @@ def context=(value)
+    # Determines whether the transaction's changeset is available for
+    # introspection.
+    #
+    # If `#buffered` is `true`, `#changes` contains the current up-to-date 
```
@bendiken this is where I think you were going with `#buffered?`. Looks good?

no-reply commented on lib/rdf/transaction.rb:
```diff
@@ -122,43 +170,21 @@ def writable?
-    def readable?
-      false # TODO: support RDF::Enumerable and RDF::Queryable.
+    # @see     RDF::Writable#mutable?
+    def mutable?
```
Depending on the outcome of the discussion at https://github.com/ruby-rdf/rdf/pull/250/files#r49671230, we may want to consider removing `writable?` and `mutable?`

no-reply commented:
I rebased this on the Hamster repository changes, implemented snapshots and atomic application of changesets, which I think completes the work.

I'm getting some `RDF::Vocabulary` test failures. These seem to be unrelated, but might be my fault due to merge conflict resolution. I'm looking into this.

In the meanwhile, feedback on this and https://github.com/ruby-rdf/rdf-spec/pull/44 is welcome. Again, I understand these to be a complete implementation of the desired interface and behavior. If something is missing, that should be called out ASAP to keep the 2.0 release on track.

cc: @gkellogg @bendiken

gkellogg commented:
See https://github.com/ruby-rdf/rdf-spec/pull/44#issuecomment-174649446. We should probably merge these PRs (modulo the `#each` implementation) and pick up the pieces.
